implement oneOf

xddq opened the pull request:
<!--
  Thanks for submitting a pull request!

  Before submitting it, please make sure that you added a test case for the bug
that you fixed, or new feature that you added. This is required.

-->

## Summary
Implements oneOf support for schema2typebox. More info [here](https://github.com/xddq/schema2typebox/issues/16)

<!--
 What did you do? Link the issue or discussion your PR is based on. Why are you
making this change?
-->

xddq commented:
Leaving this open for review **if** you want to check/comment on it in the next day or couple of days.

sinclairzx81 reviewed (approved):
@xddq Hi

The implementation of `oneOf` looks good (as per issue suggestion). Was just thinking it might be good to rename the `Kind` descriptor `OneOf` to `GeneratedOneOf` or `ExtendedOneOf` just to give users an indication that the type is an extended type (when they console.log). But this is entirely optional. Other than that, all looks good :)

### Optional Issue

Just one thing i noticed, when creating a standalone `Union`, `Intersect` or `OneOf`, that the generation / heuristics are interpreting the type as `TOptional`. For example

```typescript
// schema
{
  "anyOf": [  // same behavior for oneOf, allOf
    { "type": "string" },
    { "type": "number" },
    { "type": "boolean" }
  ]
}
// outputs
export type T = Static<typeof T>;
export const T = Type.Optional( // shouldn't be optional
  Type.Union([Type.String(), Type.Number(), Type.Boolean()])
);
```
This only seems to occur when the composite is defined at the top level, and means that if users use the generated type in a explicit `TObject` later, the property will implicitly optional.

### Automatic ID / Variable Name

Just one other thing. For schemas that include `$id`, it might be possible to use the `$id` as a variable name (instead of `T`). Just be mindful that users can specify all sorts of names for a schema $id (so will need formatting). TypeBox uses [this](https://github.com/sinclairzx81/typebox/blob/master/src/compiler/compiler.ts#L100-L113) encoding for the variable names (which is the generated function name used in the compiler). It rewrites any non-valid JS character to it's UTF-16 numeric value. 

Alternatively, a CLI option to specify the name of the type might work as well.

Other than that, all looks good. Will approve as the `oneOf` implementation looks good to me. The subsequent issues (Optional, Auto Variable Names) can probably be addressed in a separate PR / issue thread, so just making note of them here for now.

Cheers!
S

xddq commented:
@sinclairzx81 Thanks for the feedback! I will keep this up for another two days, in case you want to review the feedback as well. Otherwise I go ahead with merge and release.

## Optional Issue
great catch, fixed the implementation and the tests.

## Show user that we are using extended typebox type
replaced "OneOf" with "ExtendedOneOf"

## Automatic ID / Variable Name
did note this down, will not implement this yet, but will have a look at it once it is on my prio, thanks!
